### ISSUE TITLE
The explanation of overfit_batches is misleading

### ISSUE THREAD

**Issue by tengerye** [Lightning-AI/pytorch-lightning#19277]
### 📚 Documentation

According to the [doc](https://pytorch-lightning.readthedocs.io/en/1.0.8/debugging.html) here: 

> With this flag, the train, val, and test sets will all be the same train set. 

However, if it were true, the following snippet would not fail. Therefore, the `overfit_batches` uses both training set and validation set. I am not sure if it is the problem of it is actually a bug. Because in my understanding, overfit test is to train and test on the same dataset. In order words, my expectation is `overfit_batches` makes the same data from training dataset for both training and validation.

I would also suggest adding psudo code to help understand.

```
from torch import optim, nn, utils
import lightning as L
import torchvision.transforms as T
import torch.nn.functional as F

from torchvision import datasets
import torchvision.transforms as transforms


# define any number of nn.Modules (or use your current ones)
encoder = nn.Sequential(nn.Linear(28 * 28, 64), nn.ReLU(), nn.Linear(64, 3))
decoder = nn.Sequential(nn.Linear(3, 64), nn.ReLU(), nn.Linear(64, 28 * 28))


# define the LightningModule
class LitAutoEncoder(L.LightningModule):
    def __init__(self, encoder, decoder):
        super().__init__()
        self.encoder = encoder
        self.decoder = decoder

    def training_step(self, batch, batch_idx):
        # training_step defines the train loop.
        # it is independent of forward
        x, y = batch
        x = x.view(x.size(0), -1)
        z = self.encoder(x)
        x_hat = self.decoder(z)
        loss = nn.functional.mse_loss(x_hat, x)
        # Logging to TensorBoard (if installed) by default
        # self.print("train_loss", loss)
        return loss

    def validation_step(self, batch, batch_idx):
        # this is the test loop
        x, y = batch
        x = x.view(x.size(0), -1)
        z = self.encoder(x)
        x_hat = self.decoder(z)
        test_loss = F.mse_loss(x_hat, x)
        self.print("test_loss", test_loss)

    def configure_optimizers(self):
        optimizer = optim.Adam(self.parameters(), lr=1e-3)
        return optimizer


# init the autoencoder
autoencoder = LitAutoEncoder(encoder, decoder)


transform = transforms.ToTensor()
train_set = datasets.MNIST(root="MNIST", download=True,
                           train=True, transform=transform)
test_set = datasets.MNIST(root="MNIST", download=True, train=False, 
                          transform=T.Compose([
                                                T.Resize(512),
                                                T.ToTensor(),
                                            ])
)

train_loader = utils.data.DataLoader(train_set, 512)
val_loader = utils.data.DataLoader(test_set, 512)

# train the model (hint: here are some helpful Trainer arguments for rapid idea iteration)
trainer = L.Trainer(accelerator="gpu", devices=[
                    2], max_epochs=1, overfit_batches=1)
trainer.fit(model=autoencoder, train_dataloaders=train_loader,
            val_dataloaders=val_loader)
```

cc @borda
